### ISSUE TITLE
Androdi Talk does not establish a Phonecall connection like a browser (ends in error)

### ISSUE THREAD

**Issue by florom** [nextcloud/talk-android#1328]
Hi Devs,

I just want to bring it up here cause I am frustrated I could not get behind it.

## Steps to reproduce
1. A starts a call in Talk from PC
2. B tries to enter call in Nextcloud Android App 
3. No connection established (just spinning wheel) (On same Android with Browser connection will be established)

### Expected behaviour
Connection is established within seconds

### Actual behaviour
No connection established (just spinning wheel)

## Device information

**Device:** Motorola One
**Android version:** 10

**Talk version:** 11.1.0

## Server information

**Nextcloud version:** 21.0.2

**Talk version:** 11.2.2

**Custom Signaling server configured:** yes/no and version (see talk admin settings: `/index.php/index.php/settings/admin/talk#signaling_server`)

**Custom TURN server configured:** yes (turn only TCP only)

**Custom STUN server configured:** no (stun.nextcloud.com:443)

### Server log (data/nextcloud.log)

talkuse_pc = user which I used from PC
talkuse_android  = user which I used from android

<details>

```
[nextcloud_git.log](https://github.com/nextcloud/talk-android/files/6657916/nextcloud_git.log)

```
</details>

Aditionally from another date:
Wireshark log from firewall:

![image](https://user-images.githubusercontent.com/32240626/122106245-f7b0de80-ce19-11eb-90dd-a05550117fb0.png)

![image](https://user-images.githubusercontent.com/32240626/122106265-fd0e2900-ce19-11eb-89d9-de6c29faf408.png)

Nginx config:

server {
    listen       443 ssl;
    error_log /var/log/nginx/error.log debug;   
  
    server_name  one.example.com;
    ssl_certificate_key /root/.acme.sh/one.example.com/one.example.com.key;
    ssl_certificate     /root/.acme.sh/one.example.com/fullchain.cer;

    ssl_session_cache    shared:SSL:1m;
    ssl_session_timeout  5m;
    
    ssl_protocols  TLSv1.2;

    ssl_ciphers 'EECDH+AESGCM:EDH+AESGCM:AES256+EECDH:AES256+EDH';
    ssl_prefer_server_ciphers  on;

    location / {
        proxy_pass https://192.168.5.102;
        root   nextcloud;
        index  index.html index.htm;

        proxy_pass_header Authorization;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_http_version 1.1;
        proxy_set_header Connection "";
    proxy_buffering off;
        client_max_body_size 0;
        proxy_read_timeout 36000s;
        proxy_redirect off;
    }
}

**Comment by florom** [comment]
Hi Devs,

I wanted to update the error still exists in 12.1.1.

Hope it will geht fixed soon.

Thank you

**Comment by AndyScherzinger** [comment]
@florom Is this a general mobile app issue? -> Does it work with the iOS app?

**Comment by florom** [comment]
Hi @AndyScherzinger ,

In a few weeks I will have access to an iphone (I am not owning one) so I will give feedback.

**Comment by AndyScherzinger** [comment]
Hi @florom, that would be great. Appreciate your efforts on helping to find the cause of this issue ❤️

**Comment by florom** [comment]
Hi @AndyScherzinger ,

Thanks again for looking into this matter.

I had the chance to try it out with someone who owns an iphone 12 mini.

It is tricky.

Setting:
The overall setting stated above did not change (Updated: Nextcloud 21.0.3, Talk 11.3.1)
 I tried with the talk app only (latest version). And not with browser on the iphone.
The Endpoints where Iphone (somewhere in the world) - Firewall(IPFire,reverse proxy(nginx/letsencrypt)) - Internal Network with Nextcloud server and PC talk on the browser.

Iphone with Wlan works everytime.
Iphone with SimData(4G) never worked.

I the asked to use his old Android phone (but there is no SIM Card anymore so Wlan Only possible).
Adnroid with Wlan worked sometime.
    I got the feeling that it worked after using talk with iphone in Wlan mode.


I have wireshark logs (but not for all scenarios). If you want to take a look at them, i will anonyminze them and post them later, although I would prefer a secure channel cause I am not confident I can anonyminze wireshark logs to an extent where it is save and still useful.
In a nutshell I never saw the  row "Request UDP lifetime: 3600  ...." nevertheless it worked sometimes, which brings me back to it is tricky.

If you have questions/suggestions please ask. 

Thank you.

**Comment by mahibi** [comment]
this issue might be caused by https://github.com/nextcloud/talk-android/issues/1582

**Comment by jakobroehrl** [comment]
Maybe related: https://github.com/nextcloud/talk-android/issues/1712

**Comment by florom** [comment]
Hello all,

Short notice.
Retestet it with Nextcloud 25.0.2 and latest version of all necessary apps on 20.12.2022.
Still not working behaviour.

**Comment by mahibi** [comment]
thanks to let us know @florom , finally we are now improving the calls.
This should already be fixed on latest alpha builds, but there are some more fixes being worked on which should be available with the next release.

**Comment by xvagnet** [comment]
We have the same issue and it seems like the Talk Android app can't advertise the public IP of the mobile device to TURN server.

Scenarios we have tried (used both 4G and WiFi on mobile):
1. Mobile Firefox (no VPN) <-> Desktop Chrome = **OK**
2. Mobile Firefox (VPN) <-> Desktop Chrome = **OK**
3. Android Talk app (no VPN) <-> Desktop Chrome = **FAILS** (on chrome://webrtc-internals and on coturn logs we can see that the remote candidate for the mobile device advertises only 127.0.0.1 and not the public IP.)
4. Android Talk app (VPN) <-> Desktop Chrome = **OK** (The public VPN IP is being advertised to coturn...It seems that the VPN service takes over the advertisement of the public IP..?)

VPN is Proton and we are connecting to some random country.

We have tried TURN/STUN config in Nextcloud to be both with domain (eg. turn.example.com) and with direct IP in order to exclude DNS issues in the Talk app.

All TURN/STUN tests (inside Nextcloud admin config, online webrtc tools, coturn-utils) are OK.

Nextcloud AIO v11.7.0
Nextcloud Hub 10 (31.0.8)
Nextcloud Talk app: 21.2.0

**Comment by xvagnet** [comment]
Fixed in v22.0.0

**Comment by florom** [comment]
@xvagnet I am happy to hear about that. I am really looking forward to that. Will close issue if i can confirm.

**Comment by kami-nx** [comment]
> Nextcloud AIO v11.7.0 Nextcloud Hub 10 (31.0.8) Nextcloud Talk app: 21.2.0

I have the similar issue. Also with latest 22.0.2 Nextcloud Talk app.
Mobile phones have wireguard vpn with "don't allow traffic outside VPN" and its work like this:

NCT Mobile + VPN + WiFi = not working
NCT Mobile + VPN + LTE = not working
NCT Mobile + WiFi = working
NCT Mobile + LTE = kinda works (depend on provider, because some of them blocking some ports/traffic)

To exclude problems with VPN - the same type of VPN have on PC and Talk in browser works perfect.

Maybe there should be option: always use TURN/STUN server for connection in mobile app?

**Comment by xvagnet** [comment]
Coming back to this after HOURS of testing...
The issue here is not the Talk app - at least not the real issue.
The issue is the library it uses: libwebrtc.
In web browsers the webrtc is being handled more "intelligently" than with libwebrtc, so the configuration of the turn/stun server has to be precise. Also, on mobile networks the thing is becoming more complicated. Other things to consider are the SSL certs and IPv6.

I will not go into details now - can't really sort out all the info here in order to present them in a useful manner but feel free to ask.
I will just describe the general view of the final full working setup: 

- IPv6 is critical, since most mobile carriers provide an IPv6 along with the IPv4. The libwebrtc is stubborningly trying to use IPv6 instead of switching to IPv4 - as the "intelligent" handling of browsers do. If your VPS does not have IPv6, create a virtual one with tunnelbroker.net. Make sure your nextcloud AND your turn domains have AAAA records to point to the IPv6.

- Use eturnal: simpler and does the job. Also the logs are more meaningful. 
Config:
```
eturnal:
  realm: turn.example.com
  secret: "TURN-SECRET"
  listen:
    -
      ip: "PUBLIC_IPv4"
      port: 443
      transport: udp
      enable_turn: false
      proxy_protocol: false
    -
      ip: "PUBLIC_IPv6"
      port: 443
      transport: udp
      enable_turn: false
      proxy_protocol: false
    -
      ip: "127.0.0.1"
      port: 5349
      transport: tcp
      proxy_protocol: true

  relay_ipv4_addr: "PUBLIC_IPv4"
  relay_ipv6_addr: "PUBLIC_IPv6"
  relay_min_port: 49152
  relay_max_port: 65535

... the rest are the default or whatever you want.
```
Explanation of the above config:
-- We use 443/udp for STUN (no TURN there)
-- TURN traffic is coming in from HAProxy in 443/tcp and being redirected to a local frontend for eturnal: 127.0.0.1:5349
-- transport is tcp and not tls because the ssl termination is happening in haproxy
-- USE 443 everywhere to bypass strict networks like the mobile networks. (not sure if 443/udp for STUN makes any difference but since it's working we keep it like this. In any case, 443/tcp that goes to internal eturnal 5349/tcp has both turn/stun if the client can request it that way.)

- Which brings us to the next point: Terminate the SSL on HAProxy and DON'T use LetsEncrypt. Haproxy is better with terminating SSL and negotiating the different kinds of requests (maybe more strict when they come from libwebrtc). Also, LetsEncrypt does not seem to be accepted by libwebrtc. We used ZeroSSL.
Here is the haproxy config (on the same haproxy we handle nextcloud server, turn server and other websites):
```
frontend public                                                                                      
    bind <PUBLIC_IPv4>:443 ssl crt /etc/haproxy/certs/ alpn h2,http/1.1                                                                            
    bind <PUBLIC_IPv6>:443 ssl crt /etc/haproxy/certs/ alpn h2,http/1.1                                                                      
    mode tcp                                                                                                                                        
    tcp-request content accept if { req_ssl_hello_type 1 }                                                                           
    use_backend turn if { ssl_fc_sni -i turn.example.com }
    use_backend web if { ssl_fc_sni -i nextcloud.example.com }
    use_backend web if { ssl_fc_sni -i othersite.example.com }
    use_backend web if { ssl_fc_sni -i another.com }
    # If the client didn't provide a hostname, route it to Eturnal.
    use_backend turn if { req_ssl_sni -m len 0 }
    default_backend turn

backend turn
    mode tcp
    server turn_srv 127.0.0.1:5349 send-proxy-v2

backend web
    mode tcp
    server http_ft 127.0.0.1:4443 send-proxy-v2

frontend http_ft
  bind <PUBLIC_IPv4>:80
  bind <PUBLIC_IPv6>:80
  bind 127.0.0.1:4443 accept-proxy
  mode http
  ...
  ..
  .
```

- Talk settings in Nextcloud:
```
-- stun: turn.example.com:443
-- turns:only turn.example.com:443 <TURN-SECRET> UDP/TCP
```

All combinations of calls are now working (mobile<->browser, mobile<->mobile, with/without VPN, on 4G/5G/Wifi you name it).
The reason that it was working with VPN is that the VPN tunnel was forcing IPv4 and maybe "well-forming" the TURN/signaling requests (things that have to do with MTU and MSS Clamping, for which we also tried different configs on sysctl.conf and nftables but they seem to not matter after the above setup.)

So yes, setting up a solid TURN/STUN server is a pain - especially on mobile networks and for Android/iOS apps...
Hoping that the above info will make it less pain for anyone trying this.

**Comment by mahibi** [comment]
Thank you for your analysis @xvagnet 👍 I will keep this in mind

**Comment by florom** [comment]
@xvagnet Thanks for the analysis :)
